security/pkg/pki/custom/ca.go

```diff
@@ -0,0 +1,141 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package custom
+
+import (
+	"context"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"google.golang.org/grpc/metadata"
+
+	mesh "istio.io/api/mesh/v1alpha1"
+	pb "istio.io/api/security/v1alpha1"
+	"istio.io/istio/security/pkg/pki/signingapi"
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/env"
+	"istio.io/pkg/log"
+)
+
+const (
+	defaultRequestTimeout = 10 * time.Second
+	bearerTokenPrefix     = "Bearer "
+)
+
+var (
+	cLog = log.RegisterScope("CustomCAClient", "Custom CA Integration Log", 0)
+
+	// ClientCertDirENV Location of a local or mounted Custom CA Client TLS Certificates
+	// Includes: root-cert.pem, client.crt, client.key
+	// Users require to create kubernetes secret and mount them into this location
+	ClientCertDirENV = env.RegisterStringVar("CUSTOM_CA_CERT_DIR", "./var/lib/istio/custom-ca-tls",
+		"Location of a local or mounted Custom CA TLS certificates")
+)
+
+// CAClient generates keys and certificates for Istio identities.
+type CAClient struct {
+	signingAPIClient *signingapi.Client
+	keyCertBundle    util.KeyCertBundle
+	jwtToken         string
+}
+
+// IsUsingCustomCAForWorkloadCertificate return true if users configure CustomCA for signing Workload Certificates
+// TODO: Support for Istio Agent side
+func IsUsingCustomCAForWorkloadCertificate(caOpts *mesh.MeshConfig_CA) bool {
+	return caOpts != nil && caOpts.Address != "" && caOpts.IstiodSide
+}
+
+// NewCAClient returns a new CAClient instance.
+// CustomCAClient will use JWT TOKEN from Istiod to create Authorization header each forwarding request to External CA
+// External CA just need to verify Istiod's identity each forwarding request is sufficient
+func NewCAClient(caOpts *mesh.MeshConfig_CA, keyCertBundle util.KeyCertBundle, jwtToken string) (*CAClient, error) {
```

Comment:
Won't this break when the jwt token expires?

Reply:
I'll update code, Custom CA will forward directly JWT from Citadel to external CA